doc/decisions/DR024-Versioning-traits-and-specifications-method.md

```diff
@@ -0,0 +1,465 @@
+    successCallback(idx, success_result)
+  else:
+    raise RuntimeError()
+
+```
+
+#### Pros
+- More flexible, traits of different versions can be mixed in the same
+  session, which could allow hosts/managers to perform partial upgrades
+  in the case they are blocked on small specifics, and allow components
+  of a system to each work with a different schema version and merge
+  their results.
+- Hosts and managers would be capable of working with mismatched
+  schema versions, so long as the versions of the individual required
+  traits are compatible.
+- Hosts don't have to interrogate and declare which library versions
+  are in use, which helps especially in the case of second-order
+  dependencies.
+- Seems to fit with existing OpenAssetIO design, especially with
+  `BatchElementError`, as an incompatible version wouldn't necessarily
+  need to error the whole batch.
+- No new API, core `openassetio` library remains version unaware.
+- Cannot be ignored, unlike other "layered on top" mechanisms.
+- Managers not required to consider storage/communication of additional
+  version data to enable interpretation by future/downstream systems.
+- Stateless managers remain possible
+- Allows per-trait versioning, which avoids the need or "insider
+  knowledge" as to which traits actually change in which package
+  versions.
+- Prior art (e.g. OTIO) encodes versions in their data structure,
+  presumably for good reason.
+
+#### Cons
+- No "primary" version across the session, increasing implementation
+  complexity.
+- Would likely require some utilities in `traitgen`, otherwise, due to
+  each new version effectively being a brand new trait, constructing the
+  correct trait view would be become a tedious dance.
+- Introspection methods become more complex. (See [What's
+  versioned](#whats-versioned)).
+- Unclear how specification versioning is to work without an out-of-data
+  version.
+- The most obvious implementation, (using `traitID` to store the
+  version), would effectively make any change to a trait version a
+  breaking change. Allowing non breaking changes (such as adding a new
+  property) would require significant rework to the `traitgen` output.
+
+### In data - Extended
+The implementation of in-data versioning is very basic, which is part of
+its strength, but it necessitates a "try and see" approach on the part
+of the host. If the host wants to be adaptive to the manager, it may
+benefit from a more up front introspection of the supported manager
+versions.
+
+To this end, the `supportedSchemaVersions` method (or similar concept,)
+from Option 2 could be orthogonally added without any conceptual
+conflict. This would have the added benefit of allowing hosts to also
+support a range of version/have backwards compatibility, which is a huge
+benefit.
+
+This stops short of negotiation, being merely introspection similar to
+our other manager introspection methods.
+
+### What's versioned?
```

Comment:
[consider][minor] I think this section would be better placed before the Pros and Cons, since its introducing context for the decision.

Reply:
I don't agree, I think desciptiveness at this level of depth is better served after folks have digested the example.